packages/@aws-cdk/aws-codeartifact/lib/domain.ts

```diff
@@ -0,0 +1,223 @@
+ */
+export class Domain extends DomainBase {
+  /**
+ * Import an existing domain provided an ARN
+ *
+ * @param scope The parent creating construct
+ * @param id The construct's name
+ * @param domainArn Domain ARN (i.e. arn:aws:codeartifact:us-east-2:444455556666:domain/MyDomain)
+ */
+  public static fromDomainArn(scope: Construct, id: string, domainArn: string): IDomain {
+    const parsed = Stack.of(scope).parseArn(domainArn);
+    const domainName = parsed.resourceName || '';
+
+    class Import extends Domain {
+      public readonly domainName: string = parsed.resourceName || '';
+      public readonly domainOwner: string = parsed.account || '';
+      public readonly domainEncryptionKey: string = '';
+      public readonly domainArn = domainArn;
+    }
+
+    return new Import(scope, id, { domainName: domainName });
+  }
+
+  public readonly domainArn: string = '';
+  public readonly domainName: string = '';
+  public readonly domainOwner: string = '';
+  public readonly domainEncryptionKey: string = '' ;
+  public readonly cfnDomain: ca.CfnDomain;
+
+  constructor(scope: Construct, id: string, props: DomainProps) {
+    super(scope, id);
+
+    this.cfnDomain = new ca.CfnDomain(this, id, {
+      domainName: props.domainName,
+    });
+
+
+    if (props.domainEncryptionKey) {
+      this.cfnDomain.addPropertyOverride('EncryptionKey', props.domainEncryptionKey);
+    }
+
+    this.domainArn = this.cfnDomain.attrArn;
+    this.domainName = props.domainName;
+    this.domainOwner = this.cfnDomain.attrOwner;
+    this.domainEncryptionKey = this.cfnDomain.attrEncryptionKey;
+
+    this.Validate();
+
+    if (!props.policyDocument) {
+      const p = props.principal || new iam.AccountRootPrincipal();
+      this.allowAuthorization(p).allowCreateRepository(p).allowReadFromDomain(p);
+    } else {
+      this.cfnDomain.permissionsPolicyDocument = props.policyDocument;
+    }
+  }
+
+  private Validate() {
+    validate('DomainName', { required: true, minLength: 2, maxLength: 50, pattern: /[a-z][a-z0-9\-]{0,48}[a-z0-9]/gi }, this.domainName);
+    validate('EncryptionKey', { minLength: 1, maxLength: 2048, pattern: /\S+/gi }, this.domainEncryptionKey);
+  }
+
+  private createPolicy(principal: iam.IPrincipal, iamActions: string[], resource: string = '*') {
+    const p = this.cfnDomain.permissionsPolicyDocument as iam.PolicyDocument || new iam.PolicyDocument();
+
+    p.addStatements(new iam.PolicyStatement({
+      effect: iam.Effect.ALLOW,
+      principals: [
+        principal,
+      ],
+      resources: [resource],
+      actions: iamActions,
+    }));
+
+    this.cfnDomain.permissionsPolicyDocument = p;
+  }
+
+  /**
+     * Adds read actions for the principal to the domain's
+     * resource policy
+     * @param principal The principal for the policy
+     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/domain-policies.html
+     */
+  public allowReadFromDomain(principal : iam.IPrincipal) : Domain {
```

Comment:
`GetAuthorizationToken` should be kept as separate grant as it used in conjunction with read and write operations on repositories within the domain.  There can be a use case where you want to list repositories via the CLI but not allow the principal to interact with the repositories themselves. See here https://docs.aws.amazon.com/codeartifact/latest/ug/domain-policies.html#enabling-cross-acount-access-to-a-domain

With that said, thoughts on having a `grantAuthorization` or `grantLogin` signature?

Reply:
I suppose the dedicated `grantAuthorization` method is fine..
It then seems like the Repository's `grant*` methods should automatically call `domain.grantAuthorization`, so that if someone just grants read on a repo, they automatically get the necessary domain privileges to do so.

Reply:
I like that approach which would allow us to reduce the scope of repositories that have `GetAuthorizationToken` allowed for them. Otherwise it's wide open to allow any repository to be granted upon. 

While the `GetAuthorizationToken` does not allow permissions to manipulate or pull on a repository, it could have an adverse customer experience with the assumption that "If I'm able to get a token, then I should be able to work with the packages".